Release 0.8 seems to translate queries with ST_AsGeoJSON into wrong SQL

We have the following python code:

```python
geo_query = Session.query(
            func.ST_AsGeoJSON(func.ST_CollectionExtract(func.ST_MakeValid(RegionGeometry.geometry), 3))
                .label('geojson')).join(Shapefile.region_geometry)\
            .filter(Shapefile.id.in_(sh_id_list)).order_by(Shapefile.id).all()
```

Starting with version 0.8 of geoalchemy2 this leads to the following problem. As far as I can tell there is a missing `)` bracket before the `AS geojson` part of the query.

```
self = <sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2 object at 0x7fd8b68484e0>
cursor = <cursor object at 0x7fd8b6f08ce0; closed: -1>
statement = 'SELECT ST_AsGeoJSON(ST_CollectionExtract(ST_MakeValid(region_geometry) AS geojson \nFROM ST_CollectionExtract(ST_Make...on_geometry.region_geometry_id = shapefile.region_geometry_id \nWHERE shapefile.id IN (%(id_1)s) ORDER BY shapefile.id'
parameters = {'ST_CollectionExtract_1': 3, 'id_1': 1}
context = <sqlalchemy.dialects.postgresql.psycopg2.PGExecutionContext_psycopg2 object at 0x7fd8b60fa898>

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) syntax error at or near "AS"
E       LINE 1: ...T_CollectionExtract(ST_MakeValid(region_geometry) AS geojson...
E                                                                    ^
E       
E       [SQL: SELECT ST_AsGeoJSON(ST_CollectionExtract(ST_MakeValid(region_geometry) AS geojson 
E       FROM ST_CollectionExtract(ST_MakeValid(region_geometry.geometry), %(ST_CollectionExtract_1)s), shapefile JOIN region_geometry ON region_geometry.region_geometry_id = shapefile.region_geometry_id 
E       WHERE shapefile.id IN (%(id_1)s) ORDER BY shapefile.id]
E       [parameters: {'ST_CollectionExtract_1': 3, 'id_1': 1}]
E       (Background on this error at: http://sqlalche.me/e/f405)

```

--- a/geoalchemy2/functions.py
+++ b/geoalchemy2/functions.py
@@ -81,7 +81,9 @@ class ST_AsGeoJSON(functions.GenericFunction):
         if expr is not None:
             args = [expr] + args
         for idx, element in enumerate(args):
-            if isinstance(element, elements.HasFunction):
+            if isinstance(element, functions.GenericFunction):
+                continue
+            elif isinstance(element, elements.HasFunction):
                 if element.extended:
                     func_name = element.geom_from_extended_version
                     func_args = [element.data]
@@ -719,6 +721,9 @@ _FUNCTIONS = [
      'Returns a geometry that represents the shared portion of geomA and '
      'geomB. The geography implementation does a transform to geometry to do '
      'the intersection and then transform back to WGS84.'),
+
+    ('ST_MakeValid', types.Geometry,
+     'Attempts to make an invalid geometry valid without losing vertices.'),
 
     ('ST_LineMerge', types.Geometry,
      'Returns a (set of) LineString(s) formed by sewing together the '

--- a/tests/test_functional.py
+++ b/tests/test_functional.py
@@ -686,6 +686,14 @@ class TestCallFunction():
             "coordinates": [[0, 0], [1, 1]]
         }
 
+        # Test with function inside
+        s1_func = select([func.ST_AsGeoJSON(func.ST_MakeValid(Lake.__table__.c.geom))])
+        r1_func = session.execute(s1_func).scalar()
+        assert loads(r1_func) == {
+            "type": "LineString",
+            "coordinates": [[0, 0], [1, 1]]
+        }
+
     @skip_postgis1(postgis_version)
     @skip_postgis2(postgis_version)
     def test_ST_AsGeoJson_feature(self):
@@ -852,4 +860,15 @@ class TestSTAsGeoJson():
             ':ST_AsGeoJSON_2, :ST_AsGeoJSON_3) '
             'AS "ST_AsGeoJSON_1" FROM "another AWFUL.name for.schema".'
             '"this is.an AWFUL.name"',
+        )
+
+    @skip_postgis1(postgis_version)
+    def test_nested_funcs(self):
+        stmt = select([func.ST_AsGeoJSON(func.ST_MakeValid(func.ST_MakePoint(1, 2)))])
+        self._assert_stmt(
+            stmt,
+            'SELECT '
+            'ST_AsGeoJSON(ST_MakeValid('
+            'ST_MakePoint(:ST_MakePoint_1, :ST_MakePoint_2)'
+            ')) AS "ST_AsGeoJSON_1"',
         )

--- a/tests/test_functions.py
+++ b/tests/test_functions.py
@@ -585,6 +585,10 @@ def test_ST_Intersection():
     _test_geometry_returning_func('ST_Intersection')
 
 
+def test_ST_MakeValid():
+    _test_geometry_returning_func('ST_MakeValid')
+
+
 def test_ST_LineMerge():
     _test_geometry_returning_func('ST_LineMerge')
 

